Applications/YOLO/PyTorch/yolo_loss.py

```diff
@@ -0,0 +1,203 @@
+##
+# Copyright (C) 2023 Seungbaek Hong <sb92.hong@samsung.com>
+#
+# @file yolo_loss.py
+# @date 8 March 2023
+# @brief Define loss class for yolo
+#
+# @author Seungbaek Hong <sb92.hong@samsung.com>
+
+import torch
+import torch.nn as nn
+import torch.functional as F
+import numpy as np
+
+##
+# @brief calculate iou between two boxes list
+def calculate_iou(bbox1, bbox2):    
+    """
+    @param bbox1 shape(numb_of_bbox, 4), it contains x, y, w, h
+    @param bbox2 shape(numb_of_bbox, 4), it contains x, y, w, h
+    @return result shape(numb_of_bbox, 1)
+    """
+    # bbox coordinates
+    b1x1, b1y1 = (bbox1[:, :2]).split(1, 1)
+    b1x2, b1y2 = (bbox1[:, :2] + (bbox1[:, 2:4])).split(1, 1)
+    b2x1, b2y1 = (bbox2[:, :2]).split(1, 1)
+    b2x2, b2y2 = (bbox2[:, :2] + (bbox2[:, 2:4])).split(1, 1)
+    
+    # box areas
+    areas1 = (b1x2 - b1x1) * (b1y2 - b1y1)
+    areas2 = (b2x2 - b2x1) * (b2y2 - b2y1)
+    
+    # intersections
+    min_x_of_max_x, max_x_of_min_x = torch.min(b1x2, b2x2), torch.max(b1x1, b2x1)
+    min_y_of_max_y, max_y_of_min_y = torch.min(b1y2, b2y2), torch.max(b1y1, b2y1)
+    intersection_width = (min_x_of_max_x - max_x_of_min_x).clamp(min=0)
+    intersection_height = (min_y_of_max_y - max_y_of_min_y).clamp(min=0)
+    intersections = intersection_width * intersection_height
+    
+    # unions        
+    unions = (areas1 + areas2) - intersections
+    
+    result = intersections / unions    
+    return result
+
+##
+# @brief find best iou and its index
+def find_best_ratio(anchors, bbox):    
+    """
+    @param anchors shape(numb_of_anchors, 2), it contains w, h
+    @param bbox shape(numb_of_bbox, 2), it contains w, h
+    @return best_match index of best match, shape(numb_of_bbox, 1)
+    """
+    b1 = np.divide(anchors[:, 0], anchors[:, 1])
+    b2 = np.divide(bbox[:, 0], bbox[:, 1])
+    similarities = np.abs(b1.reshape(-1, 1) - b2)
+    best_match = np.argmin(similarities, axis=0)
+    return best_match
+
+##
+# @brief loss class for yolo
+class YoloV2_LOSS(nn.Module):
+    """Yolo v2 loss"""
```

Comment:
when you implement this loss into NNTrainer, plz consider where these computations should be done among calDerivatives, calGradient.

Reply:
thank you, I will consider your comment and work on it.